File: app/src/main/java/pemwa/com/javadevsnairobi/view/MainActivity.java
package pemwa.com.javadevsnairobi.view;

import android.app.ProgressDialog;
import android.content.Intent;
import android.os.Parcelable;
import android.support.design.widget.Snackbar;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.View;

import java.util.ArrayList;

import pemwa.com.javadevsnairobi.R;
import pemwa.com.javadevsnairobi.adapter.GithubAdapter;
import pemwa.com.javadevsnairobi.model.GithubUsers;
import pemwa.com.javadevsnairobi.presenter.GithubPresenter;
import pemwa.com.javadevsnairobi.util.NetworkUtility;

public class MainActivity extends AppCompatActivity implements UsersView, SwipeRefreshLayout.OnRefreshListener{

    private static final String TAG = "MainActivity";
    private RecyclerView recyclerView;
    private GithubAdapter adapter;
    ArrayList<GithubUsers> githubUsersArrayList;
    UsersPresenterView usersPresenterView;
    ProgressDialog progressDialog;
    GridLayoutManager gridLayoutManager;
    Parcelable parcelable;
    SwipeRefreshLayout swipeRefreshLayout;
    public static final String USERS_KEY = "users";
    public static final String RECYCLER_VIEW_STATE_KEY = "recycler_view_state";


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        swipeRefreshLayout = findViewById(R.id.swipeRefresh);
        swipeRefreshLayout.setOnRefreshListener(this);
        swipeRefreshLayout.setColorSchemeResources(
                R.color.colorPrimaryDark,
                R.color.colorPrimary,
                R.color.colorAccent);

        progressDialog = new ProgressDialog(MainActivity.this);
        progressDialog.setMessage("Loading...");
        progressDialog.setCancelable(false);
        progressDialog.show();

        recyclerView = findViewById(R.id.customRecyclerView);
        usersPresenterView = new GithubPresenter(this);
        gridLayoutManager = new GridLayoutManager(this, 2);

        if (savedInstanceState != null) {
            githubUsersArrayList = new ArrayList<>();
            githubUsersArrayList = savedInstanceState.getParcelableArrayList(USERS_KEY);
            displayGithubUsers(githubUsersArrayList);
        } else {
            loadUsers();
        }
    }

    public void loadUsers() {
        if (!(NetworkUtility.isConnected(this))) {
            progressDialog.dismiss();
            showSnackBar();
        }else {
            usersPresenterView.getGithubUsers();
        }
    }

    @Override
    protected void onResume() {
        super.onResume();

        if (parcelable != null) {
            gridLayoutManager.onRestoreInstanceState(parcelable);
        }
        Log.i(TAG, "onResume is called");
    }

    @Override
    protected void onSaveInstanceState(Bundle savedInstanceState) {
        super.onSaveInstanceState(savedInstanceState);
        savedInstanceState.putParcelableArrayList(USERS_KEY, this.githubUsersArrayList);

        parcelable = gridLayoutManager.onSaveInstanceState();
        savedInstanceState.putParcelable(RECYCLER_VIEW_STATE_KEY, parcelable);
        Log.i(TAG, "onSaveInstanceState is called");
    }

    @Override
    protected void onRestoreInstanceState(Bundle savedInstanceState) {
        super.onRestoreInstanceState(savedInstanceState);

        if (savedInstanceState != null) {
            parcelable = savedInstanceState.getParcelable(RECYCLER_VIEW_STATE_KEY);
        }
        Log.i(TAG, "onRestoreInstanceState is called");
    }

    @Override
    public void displayGithubUsers(ArrayList<GithubUsers> usersArrayList) {

        progressDialog.dismiss();
        swipeRefreshLayout.setRefreshing(false);

        githubUsersArrayList = usersArrayList;
        adapter = new GithubAdapter(usersArrayList, getApplicationContext());
        recyclerView.setAdapter(adapter);
        recyclerView.setLayoutManager(gridLayoutManager);
        adapter.notifyDataSetChanged();

        launchUserDetails();
    }

    public void launchUserDetails() {
        recyclerView.addOnItemTouchListener(new ItemClickView(getApplicationContext(), new ItemClickView.onItemClickListener() {
            @Override
            public void onItemClick(View view, int position) {
                GithubUsers user = githubUsersArrayList.get(position);

                Intent intent = new Intent(MainActivity.this, DetailActivity.class);
                Bundle bundle = new Bundle();
                bundle.putString("profPic", user.getProfilePic());
                bundle.putString("userName", user.getUserName());
                bundle.putString("url", user.getUrl());
                intent.putExtra("data", bundle);
                startActivity(intent);
            }
        }));
    }

    @Override
    public void onRefresh() {
        usersPresenterView.getGithubUsers();
    }

    private void showSnackBar() {
        Snackbar.make(
                findViewById(R.id.snackbar),
                "Failed to load users!",
                Snackbar.LENGTH_LONG)
                .setAction("TAP TO RETRY", new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        loadUsers();
                    }
                })
                .show();
    }
}

File: app/src/main/java/pemwa/com/javadevsnairobi/view/MainActivity.kt
package pemwa.com.javadevsnairobi.view

import android.app.ProgressDialog
import android.content.Intent
import android.os.Parcelable
import android.support.design.widget.Snackbar
import android.support.v4.widget.SwipeRefreshLayout
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.RecyclerView
import android.util.Log
import android.view.View

import java.util.ArrayList

import pemwa.com.javadevsnairobi.R
import pemwa.com.javadevsnairobi.adapter.GithubAdapter
import pemwa.com.javadevsnairobi.model.GithubUsers
import pemwa.com.javadevsnairobi.presenter.GithubPresenter
import pemwa.com.javadevsnairobi.util.NetworkUtility

class MainActivity : AppCompatActivity(), UsersView, SwipeRefreshLayout.OnRefreshListener {
    private var recyclerView: RecyclerView? = null
    private var adapter: GithubAdapter? = null
    private var githubUsersArrayList: ArrayList<GithubUsers>? = null
    private lateinit var usersPresenterView: UsersPresenterView
    private lateinit var progressDialog: ProgressDialog
    private lateinit var gridLayoutManager: GridLayoutManager
    private var parcelable: Parcelable? = null
    private lateinit var swipeRefreshLayout: SwipeRefreshLayout


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        swipeRefreshLayout = findViewById(R.id.swipeRefresh)
        swipeRefreshLayout.setOnRefreshListener(this)
        swipeRefreshLayout.setColorSchemeResources(
                R.color.colorPrimaryDark,
                R.color.colorPrimary,
                R.color.colorAccent)

        progressDialog = ProgressDialog(this@MainActivity)
        progressDialog.setMessage("Loading...")
        progressDialog.setCancelable(false)
        progressDialog.show()

        recyclerView = findViewById(R.id.customRecyclerView)
        usersPresenterView = GithubPresenter(this)
        gridLayoutManager = GridLayoutManager(this, 2)

        if (savedInstanceState != null) {
            githubUsersArrayList = ArrayList()
            githubUsersArrayList = savedInstanceState.getParcelableArrayList(USERS_KEY)
            displayGithubUsers((githubUsersArrayList as ArrayList<GithubUsers>?)!!)
        } else {
            loadUsers()
        }
    }

    fun loadUsers() {
        if (!NetworkUtility.isConnected(this)) {
            progressDialog.dismiss()
            showSnackBar()
        } else {
            usersPresenterView.getGithubUsers()
        }
    }

    override fun onResume() {
        super.onResume()

        if (parcelable != null) {
            gridLayoutManager.onRestoreInstanceState(parcelable)
        }
        Log.i(TAG, "onResume is called")
    }

    override fun onSaveInstanceState(savedInstanceState: Bundle) {
        super.onSaveInstanceState(savedInstanceState)
        savedInstanceState.putParcelableArrayList(USERS_KEY, this.githubUsersArrayList)

        parcelable = gridLayoutManager.onSaveInstanceState()
        savedInstanceState.putParcelable(RECYCLER_VIEW_STATE_KEY, parcelable)
        Log.i(TAG, "onSaveInstanceState is called")
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle?) {
        super.onRestoreInstanceState(savedInstanceState)

        if (savedInstanceState != null) {
            parcelable = savedInstanceState.getParcelable(RECYCLER_VIEW_STATE_KEY)
        }
        Log.i(TAG, "onRestoreInstanceState is called")
    }

    override fun displayGithubUsers(usersArrayList: ArrayList<GithubUsers>) {

        progressDialog.dismiss()
        swipeRefreshLayout.isRefreshing = false

        githubUsersArrayList = usersArrayList
        adapter = GithubAdapter(usersArrayList, applicationContext)
        recyclerView!!.adapter = adapter
        recyclerView!!.layoutManager = gridLayoutManager
        adapter!!.notifyDataSetChanged()

        launchUserDetails()
    }

    private fun launchUserDetails() {
        recyclerView!!.addOnItemTouchListener(ItemClickView(applicationContext, object : ItemClickView.onItemClickListener {
            override fun onItemClick(view: View, position: Int) {
                val user = githubUsersArrayList!![position]

                val intent = Intent(this@MainActivity, DetailActivity::class.java)
                val bundle = Bundle()
                bundle.putString("profPic", user.profilePic)
                bundle.putString("userName", user.userName)
                bundle.putString("url", user.url)
                intent.putExtra("data", bundle)
                return startActivity(intent)
            }
        }))
    }

    override fun onRefresh() {
        usersPresenterView.getGithubUsers()
    }

    private fun showSnackBar() {
        Snackbar.make(
                findViewById(R.id.snackbar),
                "Failed to load users!",
                Snackbar.LENGTH_LONG)
                .setAction("TAP TO RETRY") { loadUsers() }
                .show()
    }

    companion object {

        private val TAG = "MainActivity"
        val USERS_KEY = "users"
        val RECYCLER_VIEW_STATE_KEY = "recycler_view_state"
    }
}
